### PR TITLE
fix(schema): use one cache entree for tenant path schemas

### LINKED ISSUE #1014: tree items with many children hang browser
https://nda.ya.ru/t/D4Zkbcn176xy6F

Image

### PR DESCRIPTION
**Pull request by ValeraS**
Closes #1014

### CODE REVIEW

File: src/containers/Tenant/Diagnostics/HotKeys/HotKeys.tsx [hunk 1]
```diff
@@ -65,7 +65,7 @@ export function HotKeys({path}: HotKeysProps) {
         schemaApi.endpoints.getSchema.useQueryState({path});
     const schemaLoading = schemaIsFetching && schemaData === undefined;
 
-    const keyColumnsIds = schemaData?.PathDescription?.Table?.KeyColumnNames;
+    const keyColumnsIds = schemaData?.[path]?.PathDescription?.Table?.KeyColumnNames;
```
Comment (artemmufazalov) [inline src/containers/Tenant/Diagnostics/HotKeys/HotKeys.tsx:68]: Schema tab displays nothing

Screenshot 2024-07-15 at 16 29 38

File: src/store/reducers/schema/schema.ts [hunk 2]
```diff
@@ -51,61 +51,60 @@ export const schemaApi = api.injectEndpoints({
...
             keepUnusedDataFor: Infinity,
-            forceRefetch: ({endpointState}) => {
-                const data = endpointState?.data;
-                if (data && typeof data === 'object' && 'partial' in data && data.partial) {
-                    return true;
-                }
-                return false;
+            serializeQueryArgs: ({queryArgs: {path}}) => {
+                const parts = path.split('/');
+                return {path: parts[0] || parts[1]};
```
Comment (artemmufazalov) [inline src/store/reducers/schema/schema.ts:66]: How does this serialize works? Data is cached for the tenant?

**Review by artemmufazalov — CHANGES_REQUESTED** [review]

File: src/store/reducers/schema/schema.ts (hunk 2 above)
Comment (ValeraS) [reply src/store/reducers/schema/schema.ts:71]: Yes

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 3]
```diff
@@ -33,15 +33,26 @@ export function SchemaTree(props: SchemaTreeProps) {
...
+            const {data, originalArgs} = await promise;
+            promise.unsubscribe();
+            // Check if the result from the current request is received. rtk-query may skip the current request and
+            // return data from a parallel request, due to the same cache key.
+            if (originalArgs?.path === path) {
+                schemaData = data?.[path];
+                break;
+            }
+            // eslint-disable-next-line no-constant-condition
+        } while (true);
```
Comment (artemmufazalov) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:50]: Why to do this in `do while` cycle? `await` is not enough?

**Review by artemmufazalov — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx (hunk 3 above)
Comment (ValeraS) [reply src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:50]: As described in the comment, the "promise" may return result from another request (with another params), so we try to repeat the request in this case.

**Review by artemmufazalov — APPROVED** [review]

**Pull request merged**